framework/src/org/checkerframework/framework/stub/StubParser.java

```diff
@@ -841,111 +854,95 @@ private ClassOrInterfaceType unwrapDeclaredType(Type type) {
+                        annotate(
+                                adeclType.getTypeArguments().get(i),
+                                declType.getTypeArguments().get().get(i),
+                                null);
+                    }
+                }
+                break;
+            case WILDCARD:
+                AnnotatedWildcardType wildcardType = (AnnotatedWildcardType) atype;
+                WildcardType wildcardDef = (WildcardType) typeDef;
+                if (wildcardDef.getExtendedType().isPresent()) {
+                    annotate(
+                            wildcardType.getExtendsBound(),
+                            wildcardDef.getExtendedType().get(),
+                            null);
+                    annotate(wildcardType.getSuperBound(), primaryAnnotations);
+                } else if (wildcardDef.getSuperType().isPresent()) {
+                    annotate(wildcardType.getSuperBound(), wildcardDef.getSuperType().get(), null);
+                    annotate(wildcardType.getExtendsBound(), primaryAnnotations);
                 } else {
-                    // Duplicate parameter annotations to the type.
-                    param.getType().setAnnotations(param.getAnnotations());
+                    annotate(atype, primaryAnnotations);
                 }
-                annotate(paramType, param.getType());
-            }
-        }
-
-        if (decl.getReceiverParameter().isPresent()
-                && !decl.getReceiverParameter().get().getAnnotations().isEmpty()) {
-            if (methodType.getReceiverType() == null) {
-                stubAlwaysWarn(
-                        String.format(
-                                "parseConstructor: constructor of a top-level class cannot have receiver annotations%n"
-                                        + "Constructor: %s%n"
-                                        + "Receiver annotations: %s",
-                                methodType, decl.getReceiverParameter().get().getAnnotations()));
-            } else {
-                annotate(
-                        methodType.getReceiverType(),
-                        decl.getReceiverParameter().get().getAnnotations());
-            }
+                break;
+            case TYPEVAR:
+                // Add annotations from the declaration of the TypeVariable
+                AnnotatedTypeVariable typeVarUse = (AnnotatedTypeVariable) atype;
+                for (AnnotatedTypeVariable typePar : typeParameters) {
+                    if (typePar.getUnderlyingType() == atype.getUnderlyingType()) {
+                        AnnotatedTypeMerger.merge(
+                                typePar.getUpperBound(), typeVarUse.getUpperBound());
+                        AnnotatedTypeMerger.merge(
+                                typePar.getLowerBound(), typeVarUse.getLowerBound());
+                    }
+                }
+                break;
+            default:
+                // No annotations to add.
```

Comment:
Maybe `No further annotations to add.` would make it more explicit that the primary annotation was already added earlier.